Implement multipartI modifier

fluffynukeit opened the pull request:
This branch implements multipart input to go along with the multipart output combinator.  I have tested it with curl and it seems to work fine.  There also looks to be no impact on my json GET/POST resources, so I don't think it breaks anything.  I did not test or assess the impact on rest-gen for docs/clients.

I have the following concerns about my implementation.  Not sure how problematic they might be.
- The multipart boundary is carried around in the MultipartFormat data contructor.  This makes Format no longer an Enum or a Bounded.  
- The BodyPart type is not re-exported from Rest.Dictionary.Combinators, so using mutlipartI requires importing Network.Multipart to get access to the type.  I don't know what the usual practice is for this kind of thing, so I left it alone.
- In this implementation, the body parts are all read as bytestrings and it is up to the handler to inspect headers in the BodyParts and parse the bytestrings as appropriate.  This requires either very complicated handling code or some kind of convention on the endpoint for significance and order of BodyParts (for instance, in my test code I check the number of body parts and throw UnsupportedFormat if the number is wrong).  I suspect there might be a smarter design that puts the header inspection/parsing into the backend code and out of the handler.

Thanks!  I am really loving the rest library.

fluffynukeit committed:
Add multipartI modifier.

Formatting updates requested by @hesselink

bergmark commented:
Is this still a good base for getting this implemented?

hesselink commented:
I think so, if someone fixes the comments above this could still work. I'm still working on the dictionary rewrite, but that's nowhere near finished.

jcpetruzza committed:
Fix typos in haddock for Param dictionary

fluffynukeit commented:
Hey, gents.  I'm pretty far removed from this PR now doing other things, but at the very least I can fix up all the cosmetic nits above, and optionally rebase onto a branch of your choosing.  The fix to use multipart instead of the broken string parsing will probably need to come from you guys, though.  Does that work for you both?

bergmark commented:
Thanks for letting us know Daniel!

Any help is much appreciated, rebasing is best made on top of master.

fluffynukeit commented:
Ok, I implemented all the formatting and style updates, I think.  I rebased on master.  My build environment changed a lot so I was actually unable to build and test the changes, but I suspect they are ok and any required fixes to be minor.  Let me know if there's anything else I can help with.

hesselink commented:
@fluffynukeit Thanks!

@bergmark Shall I look into the parseContentType/multipart stuff, or will you?

bergmark commented:
@hesselink if you have time I'd appreciate it! If we're able to merge this soonish we can include it in the pending major release.

hesselink commented:
I'll try to have a look tonight.

bergmark commented:
@hesselink any chance you'll be able to take a look at this soon? I think everything else is ready for release now.

hesselink commented:
Sorry, forgot to reply to this one. Turns out I'm pretty busy :( Feel free to do it ;)

bergmark commented:
Ok, I think I'll just release what we have now. I'm eager to get it out the door!
